### DISCUSSION TITLE
generic header passing through decorator of view functions

### DISCUSSION THREAD

**Discussion by drake-mer** [fastapi/fastapi#6421] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# you might need pytest to execute the test
# (you can also run the function directly)

import functools

from typing import Optional

from fastapi import FastAPI
from fastapi.params import Header
from fastapi.testclient import TestClient

app = FastAPI()

client = TestClient(app)

headers_passed = []


def header_processor(f):
    @functools.wraps(f)
    def wraps(*args, x_custom_header: Optional[bool] = Header(False), **kwargs):
        headers_passed.append(x_custom_header)
        return f(*args, **kwargs)

    return wraps


@app.get("/")
@header_processor
def read_root(
    # here it does not work because we do not explicitely specify
    # the header as a keyword argument for the controllers definition
):
    return {"Hello": "World"}


@app.get("/items/{item_id}")
@header_processor
def read_item(
    item_id: int,
    q: Optional[str] = None,
    # here it works
    x_custom_header: Optional[bool] = Header(None),
):
    return {"item_id": item_id, "q": q}


def test_header_pass_through():
    client.get("/", headers={"x-custom-header": "true"})
    client.get("/items/2", headers={"x-custom-header": "true"})
    assert len(headers_passed) == 2
    assert headers_passed[1] is True
    assert headers_passed[0] is True

if __name__ == '__main__':
    test_header_pass_through()
```

### Description

You can run the above code and verify that it does not work.

It is a bit strange, because I assumed that decorating the function would have the effect of adding a header parser to every request of this application.

If you have a simple explanation for that, I would be glad, and could eventually be able to submit a pull request if the behavior highlighted here is not a deliberate design choice.

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

Python 3.9.0

**Comment by ghandic** [comment]
I believe FastAPI uses the function signature, so your wrapper function should instead modify the signature adding an extra kwarg on using the inspect module

**Comment by drake-mer** [comment]
I realized that just after posting, and so I checked at the python documentation, but it seems annotations are actually updated by `functools.wraps`.
But indeed, I might need to dig this possibility in more depth.
 Thanks for the comment !

**Comment by ghandic** [comment]
I do a similar thing in my fastapi deepobject implementation where I modify the signature. Can check it out for reference and implement it as a decorator instead

**Comment by uriyyo** [comment, marked as answer]
@dkremer-ledger Could you please try to replace `header_processor` with:
```py
import inspect

def header_processor(f):
    def wrapper(*args, x_custom_header: Optional[bool] = Header(False), **kwargs):
        headers_passed.append(x_custom_header)
        return f(*args, **kwargs)

    sign = inspect.signature(f)

    if "x_custom_header" not in sign.parameters:
        wrapper_sign = inspect.signature(wrapper)

        sign = sign.replace(
            parameters=[
                *sign.parameters.values(),
                wrapper_sign.parameters["x_custom_header"],
            ],
        )

    functools.update_wrapper(wrapper, f)
    wrapper.__signature__ = sign
    return wrapper
```

**Comment by drake-mer** [comment]
@uriyyo It just works 
I think it could be worth to mention it in the documentation, looks like an elegant way to specify a header dependency in a fast api project.
